Sources/Plasma/NucleusLib/pnAsyncCoreExe/pnAceSocket.cpp

```diff
@@ -0,0 +1,543 @@
+    tcp::socket             fSock;
+    plNetAddress            fRemoteAddr;
+    AsyncCancelId           fCancelId;
+    FAsyncNotifySocketProc  fNotifyProc;
+    void*                   fParam;
+    std::vector<uint8_t>    fConnectBuffer;
+    unsigned int            fConnectionType;
+
+    ConnectOperation(asio::io_context& context)
+        : fSock(context), fCancelId(), fParam(), fConnectionType()
+    { }
+    ConnectOperation(const ConnectOperation&) = delete;
+};
+
+struct WriteOperation
+{
+    size_t                  fAllocSize;
+    AsyncNotifySocketWrite  fNotify;
+    unsigned                queueTimeMs;
+
+    WriteOperation() : fAllocSize() { }
+
+    asio::const_buffer AsBuffer() const
+    {
+        return asio::buffer(fNotify.buffer + fNotify.bytesCommitted, fNotify.bytes - fNotify.bytesCommitted);
+    }
+};
+
+struct AsyncSocketStruct
+{
+    std::recursive_mutex        fCritsect;
+    tcp::socket                 fSock;
+    FAsyncNotifySocketProc      fNotifyProc;
+    void*                       fParam;
+    unsigned int                fConnectionType;
+    void*                       fUserState;
+    uint8_t                     fBuffer[kAsyncSocketBufferSize];
+    size_t                      fBytesLeft;
+    std::list<WriteOperation *> fWriteOps;
+    unsigned                    initTimeMs;
+    unsigned                    closeTimeMs;
+
+    AsyncSocketStruct(ConnectOperation& op)
+        : fSock(std::move(op.fSock)), fNotifyProc(op.fNotifyProc),
+          fParam(op.fParam), fConnectionType(op.fConnectionType),
+          fUserState(), fBuffer(), fBytesLeft(), initTimeMs(), closeTimeMs()
+    { }
+};
+
+static AsyncIoPool*                 s_ioPool;
+
+static std::recursive_mutex         s_connectCrit;
+static std::list<ConnectOperation>  s_connectList;
+static uintptr_t                    s_nextConnectCancelId = 1;
+
+void SocketInitialize()
+{
+    if (s_ioPool)
+        return;
+
+    s_ioPool = new AsyncIoPool;
+}
+
+void SocketDestroy(unsigned exitThreadWaitMs)
+{
+    // Cancel any pending connect requests
+    {
+        hsLockGuard(s_connectCrit);
+        for (ConnectOperation& op : s_connectList) {
+            asio::error_code err;
+            op.fSock.cancel(err);  // This will wake up async_connect()
+            if (err)
+                LogMsg(kLogError, "Failed to cancel socket connect operation: {}", err.message());
+        }
+    }
+
+    if (s_ioPool) {
+        s_ioPool->Destroy(exitThreadWaitMs);
+        delete s_ioPool;
+        s_ioPool = nullptr;
+    }
+}
+
+static void SocketStartAsyncRead(AsyncSocket sock)
+{
+    hsLockGuard(sock->fCritsect);
+    uint8_t* start = sock->fBuffer + sock->fBytesLeft;
+    size_t   count = sizeof(sock->fBuffer) - sock->fBytesLeft;
+    sock->fSock.async_read_some(asio::buffer(start, count),
+                                [sock](const asio::error_code& err, size_t bytes) {
+        if (err) {
+            if (err.value() == asio::error::operation_aborted) {

```

Comment:
There's a gap here that caused the deadlock when the UUID was wrong on the Mac client. When the server disconnects, we get an EOF error. But EOFs aren't handled here. My thought is an error should be handled like a disconnect?

Reply:
I would consider an EOF a disconnect.

Reply:
It's in. plFilePatcher still deadlocks when the Cyan Worlds Gatekeeper hangs up on it, but the network code is at least still routing the message out correctly. I think somewhere deeper in the code there is a while loop that doesn't do error checking. Likely beyond the scope of this PR.

Reply:
That seems likely. We had to add special handling for Cyan's hangup to the Windows login dialog.